geo/src/algorithm/dimensions.rs

```diff
@@ -0,0 +1,452 @@
+    }
+}
+
+impl<C: CoordinateType> HasDimensions for Line<C> {
+    fn is_empty(&self) -> bool {
+        false
+    }
+
+    fn dimensions(&self) -> Dimensions {
+        if self.start == self.end {
+            // degenerate line is a point
+            Dimensions::ZeroDimensional
+        } else {
+            Dimensions::OneDimensional
+        }
+    }
+
+    fn boundary_dimensions(&self) -> Dimensions {
+        if self.start == self.end {
+            // degenerate line is a point, which has no boundary
+            Dimensions::Empty
+        } else {
+            Dimensions::ZeroDimensional
+        }
+    }
+}
+
+impl<C: CoordinateType> HasDimensions for LineString<C> {
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn dimensions(&self) -> Dimensions {
+        if self.0.is_empty() {
+            return Dimensions::Empty;
+        }
+
+        debug_assert!(self.0.len() > 1, "invalid line_string with 1 coord");
+        let first = self.0[0];
+        if self.0.iter().any(|&coord| first != coord) {
+            Dimensions::OneDimensional
+        } else {
+            // all coords are the same - i.e. a point
+            Dimensions::ZeroDimensional
+        }
+    }
+
+    /// ```
+    /// use geo_types::line_string;
+    /// use geo::algorithm::dimensions::{HasDimensions, Dimensions};
+    ///
+    /// let ls = line_string![(x: 0.,  y: 0.), (x: 0., y: 1.), (x: 1., y: 1.)];
+    /// assert_eq!(Dimensions::ZeroDimensional, ls.boundary_dimensions());
+    ///
+    /// let ls = line_string![(x: 0.,  y: 0.), (x: 0., y: 1.), (x: 1., y: 1.), (x: 0., y: 0.)];
+    /// assert_eq!(Dimensions::Empty, ls.boundary_dimensions());
+    ///```
+    fn boundary_dimensions(&self) -> Dimensions {
+        if self.is_closed() {
+            return Dimensions::Empty;
+        }
+
+        match self.dimensions() {
+            Dimensions::Empty | Dimensions::ZeroDimensional => Dimensions::Empty,
+            Dimensions::OneDimensional => Dimensions::ZeroDimensional,
+            Dimensions::TwoDimensional => unreachable!("line_string cannot be 2 dimensional"),
+        }
+    }
+}
+
+impl<C: CoordinateType> HasDimensions for Polygon<C> {
+    fn is_empty(&self) -> bool {
+        self.exterior().is_empty()
+    }
+
+    fn dimensions(&self) -> Dimensions {
+        Dimensions::TwoDimensional
+    }
+
+    fn boundary_dimensions(&self) -> Dimensions {
+        Dimensions::OneDimensional
+    }
+}
+
+impl<C: CoordinateType> HasDimensions for MultiPoint<C> {
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn dimensions(&self) -> Dimensions {
+        if self.0.is_empty() {
+            return Dimensions::Empty;
+        }
+
+        Dimensions::OneDimensional
+    }
+
+    fn boundary_dimensions(&self) -> Dimensions {
+        Dimensions::Empty
+    }
+}
+
+impl<C: CoordinateType> HasDimensions for MultiLineString<C> {
+    fn is_empty(&self) -> bool {
+        if self.0.is_empty() {
```

Comment:
Why is this needed?  Iterator::all returns true on empty iterator (Iterator::any returns false).

Reply:
Good point - removed!